samples/dnn/speech_recognition.cpp

```diff
@@ -0,0 +1,581 @@
+            }
+            x_std /= elments_num;
+            x_std = sqrt(x_std) + 1e-10; // make sure x_std is not zero
+
+            for(size_t j = 0; j < elments_num; ++j)
+            {
+                x_stft[i][j] = (x_stft[i][j] - x_mean) / x_std; // standard score
+            }
+        }
+        return x_stft;
+    }
+};
+
+class Decoder {
+    // Used for decoding the output of jasper model
+private:
+    unordered_map<int, char> labels_map = fillMap();
+    int blank_id = 28;
+
+public:
+    unordered_map<int, char> fillMap()
+    {
+        vector<char> labels={' ','a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p'
+                                ,'q','r','s','t','u','v','w','x','y','z','\''};
+        unordered_map<int, char> map;
+        for(int i = 0; i < static_cast<int>(labels.size()); ++i)
+        {
+            map[i] = labels[i];
+        }
+        return map;
+    }
+
+    string decode(Mat& x)
+    {
+        // Takes output of Jasper model and performs ctc decoding algorithm to
+        // remove duplicates and special symbol. Returns prediction
+
+        vector<int> prediction;
+        for(int i = 0; i < x.size[1]; ++i)
+        {
+            double maxEl = -1e10;
+            int ind = 0;
+            for(int j = 0; j < x.size[2]; ++j)
+            {
+                if (maxEl <= x.at<float>(0, i, j))
+                {
+                    maxEl = x.at<float>(0, i, j);
+                    ind = j;
+                }
+            }
+            prediction.push_back(ind);
+        }
+        // CTC decoding procedure
+        vector<double> decoded_prediction = {};
+        int previous = blank_id;
+
+        for(int i = 0; i < static_cast<int>(prediction.size()); ++i)
+        {
+            if (( prediction[i] != previous || previous == blank_id) &&  prediction[i] != blank_id)
+            {
+                decoded_prediction.push_back(prediction[i]);
+            }
+            previous = prediction[i];
+        }
+
+        string hypotheses = {};
+        for(size_t i = 0; i < decoded_prediction.size(); ++i)
+        {
+            auto it = labels_map.find(static_cast<char>(decoded_prediction[i]));
+            if (it != labels_map.end())
+                hypotheses.push_back(it->second);
+        }
+        return hypotheses;
+    }
+
+};
+
+static string predict(std::vector<std::vector<double>>& features, dnn::Net net, Decoder decoder)
+{
+    // Passes the features through the Jasper model and decodes the output to english transcripts.
+
+    // expand 2d features matrix to 3d Mat
+    std::vector<int> sizes = {1, static_cast<int>(features.size()),
+                              static_cast<int>(features[0].size())};
+    Mat featuresMat(sizes, CV_64F);
+    for(int i = 0; i < featuresMat.size[1]; ++i)
+    {
+        for(int j = 0; j < featuresMat.size[2]; ++j)
+        {
+            featuresMat.at<double>(0, i, j) = features[i][j];
+        }
+    }
```

Comment:
> std::vector<std::vector<double>>& features

Please use `Mat` in `calculate_features()` instead and remove such manual "copy" conversion code (multiple places).

---

Use "copy-free" `.reshape()" to convert 2D into 3D tensor.